cpp/src/arrow/compute/exec/asof_join_benchmark.cc

```diff
@@ -0,0 +1,456 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <boost/process.hpp>
+#include <string>
+
+#include "benchmark/benchmark.h"
+
+#include "arrow/compute/exec/test_util.h"
+#include "arrow/csv/writer.h"
+#include "arrow/dataset/file_parquet.h"
+#include "arrow/filesystem/api.h"
+#include "arrow/ipc/api.h"
+#include "arrow/table.h"
+#include "arrow/testing/future_util.h"
+
+namespace arrow {
+namespace compute {
+
+static const char* time_col = "time";
+static const char* key_col = "id";
+static bool createdBenchmarkFiles = false;
+static std::shared_ptr<arrow::internal::TemporaryDir> temp_dir =
+arrow::internal::TemporaryDir::Make("asof_benchmarks").ValueOrDie();
+
+struct ReaderNodeTableProperties {
+  ExecNode* execNode;
+  size_t total_rows;
+  size_t total_bytes;
+};
+
+// requires export PYTHONPATH=/path/to/benchmark_scripts/table_generation
+// calls generate_benchmark_files to create tables (feather files) varying in frequency,
+// width, key density for benchmarks. places generated files in benchmark/data. This
+// operation runs once at the beginning of benchmarking.
+static void DoSetup() {
+  if (!createdBenchmarkFiles) {
+    std::error_code err;
+    std::ostringstream call_stream;
+    call_stream << "python3 -m generate_benchmark_files " << temp_dir.get()->path().ToString();
+    boost::process::system(call_stream.str(), err);
+    if (err) {
+      std::cerr << "Could not generate python files." << std::endl;
+      std::cerr << "Error Message: " << err.message() << std::endl;
+    }
+    createdBenchmarkFiles = true;
+  }
+}
+
+static std::vector<std::string> generateRightHandTables(std::string freq, int width_index,
+                                                        int num_tables,
+                                                        int num_ids_index) {
+  auto const generate_file_name = [](std::string freq, std::string is_wide,
+                                     std::string num_ids, std::string num) {
+    return freq + "_" + is_wide + "_" + num_ids + num + ".feather";
+  };
+
+  std::string width_table[] = {"20_cols", "100_cols", "500_cols"};
+  std::string num_ids_table[] = {"100_ids", "5000_ids", "10000_ids"};
+
+  std::vector<std::string> right_hand_tables;
+  right_hand_tables.reserve(num_tables);
+
+  for (int j = 1; j <= num_tables; j++) {
+    right_hand_tables.push_back(generate_file_name(
+        freq, width_table[width_index], num_ids_table[num_ids_index], std::to_string(j)));
+  }
+  return right_hand_tables;
+}
+
+// Wrapper to enable the use of RecordBatchFileReaders as RecordBatchReaders
+class RecordBatchFileReaderWrapper : public arrow::ipc::RecordBatchReader {
+  std::shared_ptr<arrow::ipc::RecordBatchFileReader> _reader;
+  int _next;
```

Comment:
```suggestion
  std::shared_ptr<arrow::ipc::RecordBatchFileReader> reader_;
  int next_;
```

Perhaps `next_row_group_` or `next_batch_index_`?  But `next_` is ok too.